File: OPP003/example-sparse.f90
! OPP003: Offloading Opportunity 
! https://www.codee.com/knowledge/opp003/
! Sparse reduction pattern
! https://www.codee.com/knowledge/patterns/sparse-reduction/

SUBROUTINE example(A, nodes1, nodes2, nodes3, n)
    integer, intent(in) :: n
    integer, dimension (1:n), intent(in) :: nodes1, nodes2, nodes3
    double precision, dimension (1:n), intent(out) :: A
    integer :: nel

    DO nel=1,n
       A(nodes1(nel)) = A(nodes1(nel)) + (nel * 1)
    END DO 
END

File: OPP003/example-sparse.cpp
// OPP003: Offloading Opportunity
// https://www.codee.com/knowledge/opp003/
// Sparse reduction pattern
// https://www.codee.com/knowledge/patterns/sparse-reduction/

void example(double *A, int *nodes1, int *nodes2, int *nodes3, int n) {
  for (int nel = 0; nel < n; ++nel) {
    A[nodes1[nel]] += nel * 1;
  }
}
